geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/SortingSimpleFeatureIterator.scala

```diff
@@ -0,0 +1,289 @@
+   * @param ordering feature ordering
+   */
+  private class MergeSortingIterator(
+      files: IndexedSeq[File],
+      serializer: SimpleFeatureSerializer,
+      mem: Iterator[SimpleFeature],
+      closeable: CloseableIterator[_],
+      ordering: Ordering[SimpleFeature]
+    ) extends CloseableIterator[SimpleFeature] {
+
+    // list of feature iterators to merge
+    private val merging = IndexedSeq(CloseableIterator(mem)) ++ files.map(new FileIterator(_, serializer))
+
+    // priority queue containing the next feature from each file
+    private val heads = {
+      val o = Ordering.by[(SimpleFeature, Int), SimpleFeature](_._1)(ordering)
+      val q = new java.util.PriorityQueue[(SimpleFeature, Int)](files.size + 1, o)
+      var i = 0
+      while (i < merging.length) {
+        val m = merging(i)
+        if (m.hasNext) {
+          q.add(m.next -> i)
+        }
+        i += 1
+      }
+      q
+    }
+
+    override def hasNext: Boolean = !heads.isEmpty
+
+    override def next(): SimpleFeature = {
+      val (f, i) = heads.poll()
+      val m = merging(i)
+      if (m.hasNext) {
+        heads.add(m.next -> i)
+      }
+      f
+    }
+
+    override def close(): Unit = CloseQuietly.raise(merging :+ closeable)
+  }
+
+  /**
+   * Reads a tmp file of serialized simple features
+   *
+   * @param file file
+   * @param serializer serializer
+   */
+  private class FileIterator(file: File, serializer: SimpleFeatureSerializer)
+      extends CloseableIterator[SimpleFeature] with LazyLogging {
+
+    private val is = new FileInputStream(file)
+    private var buf = Array.ofDim[Byte](4)
+
+    override def hasNext: Boolean = is.available() > 0
+
+    override def next(): SimpleFeature = {
+      is.read(buf, 0, 4)
+      val size = ByteArrays.readInt(buf)
+      if (buf.length < size) {
+        buf = Array.ofDim((size * 1.2).toInt)
+      }
+      is.read(buf, 0, size)
+      serializer.deserialize(buf, 0, size)
+    }
+
+    override def close(): Unit = {
+      try { is.close() } finally {
+        if (!file.delete()) {
+          logger.warn(s"Unable to delete tmp file '${file.getAbsolutePath}''")
+          file.deleteOnExit()
+        }
+      }
+    }
+  }
+
+  /**
+   * Evidence that a feature is sizable
+   *
+   * @tparam T type binding
+   */
+  private trait FeatureIsSizable[T <: SimpleFeature] {
+    def sizeOf(feature: T): Long
+  }
+
+  /**
+   * Evidence that Sizable features are sizable
+   */
+  private object SizableFeatureIsSizable extends FeatureIsSizable[SimpleFeature with Sizable] {
+    override def sizeOf(feature: SimpleFeature with Sizable): Long = feature.calculateSizeOf()
+  }
+
+  /**
+   * Use a rough estimate based on attributes and user data
+   */
+  private object UnSizableFeatureIsSizable extends FeatureIsSizable[SimpleFeature] {
+    override def sizeOf(feature: SimpleFeature): Long =
+      Sizable.sizeOf(feature) + Sizable.deepSizeOf(feature.getAttributes, feature.getUserData)
```

Comment:
In this case, do we expect Sizable.sizeOf(feature) to return that 64 as a default value?

Reply:
only if the actual sizeof agent can't be loaded. deepSizeOf delegates to the sizeof function but has some additional logic, so I'm not sure exactly what it would return